### PR TITLE
Add explicit braces, C++ casts, turn on warnings with compile

### PR DESCRIPTION
**Pull request by Veratil**
* Bump minimum CMake version to 3.1, noticed this is minimum for `CMAKE_CXX_STANDARD`
* Add the `WITH_WARNINGS` build option I've been using for testing various warning flags (using `/W4` for MSVC since I saw somewhere that's a similar level for `-Wall -Wextra`)
* Add explicit braces where missing, pretty sure I got them all
* Update to C++ style casts as best as I could, there's a lot of C style casting that just doesn't fly well with C++

The `WITH_WARNINGS` I made work with gcc and clang (including macOS clang) and threw in the MSVC branch for when we can test that better.

Apologies for the braces and C++ casts together in the same commit. I was going through all lines by hand and just did it at the same time.

### CODE REVIEW

**Comment by Veratil** [comment]
Hold off on merge for now. I want to get a better test done with this build first. I can do that tomorrow. 👍

**Comment by Veratil** [comment]
Alright I've tested this build with LMMS (embedded not library yet). MIDI imports are still working as before. I'm confident about the casts. 👍

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,18 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.1)
 
 project(PortSMF VERSION 235.1 LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 11)
 
+option(WITH_WARNINGS "Build with compiler warnings enabled" OFF)
+if (WITH_WARNINGS)
```
Comment (Be-ing) [inline CMakeLists.txt:7]:
```suggestion
if(WITH_WARNINGS)
```
for consistency

Comment (Be-ing) [reply CMakeLists.txt:7]: also change below

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: include/allegro.h [hunk 2]
```diff
@@ -513,7 +520,9 @@ typedef class Alg_time_map {
     // tempo change takes effect before the inserted beats
     void insert_beats(double start, double len);
     void dereference() {
-        if (--refcount <= 0) delete this;
+        if (--refcount <= 0) {
```
Comment (Be-ing) [inline include/allegro.h:523]: aaa more mutating variables in conditional expressions. Please separate this from the conditional expression

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,18 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.1)
 
 project(PortSMF VERSION 235.1 LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 11)
 
+option(WITH_WARNINGS "Build with compiler warnings enabled" OFF)
```
Comment (Be-ing) [inline CMakeLists.txt:6]: If it builds without warnings, let's not make this an option and just always turn on warnings.

**Comment by Be-ing** [comment]
Thank you for doing this tedious cleanup.

File: CMakeLists.txt (hunk 3 above)
Comment (Veratil) [reply CMakeLists.txt]: Done!

File: include/allegro.h (hunk 2 above)
Comment (Veratil) [reply include/allegro.h]: oops it inserted a tab instead of spaces again :joy: gotta fix that in my vim config lol

**Comment by Veratil** [comment]
Okay! Now it all should be good. 😂

**Comment by Veratil** [comment]
Oh I should probably rename the first commit

**Comment by Veratil** [comment]
Lots of warnings in MSVC, but nothing that's breaking.

**Comment by Be-ing** [comment]
Let me know when you're done with the cleanups you want to do and I'll tag a release.

**Comment by Veratil** [comment]
Currently working on adding Doxygen. It won't be *good* initial documentation since it's kind of "these groups of functions do this" or there just isn't any, but it'll be a stepping point.

**Comment by Be-ing** [comment]
@Veratil my [PR to package PortSMF for upstream vcpkg](https://github.com/microsoft/vcpkg/pull/18982) was merged.

**Pull request merged**